core(image-size-responsive): quantize DPRs

sk- opened the pull request:
<!-- Thank you for submitting a pull request! -->

**Summary**
<!-- What kind of change does this PR introduce? -->
<!-- Is this a bugfix, feature, refactoring, build related change, etc? -->
<!-- Describe the need for this change -->
In some contexts lighthouse uses a DPR of 2.625, which is a rather uncommon DPR, so to minimize the number of failures, we will quantize the original DPR to the lowest integer (or 1.5 if in between 1.5 and 2).

<!-- Link any documentation or information that would help understand this change -->

**Related Issues/PRs**
<!-- Provide any additional information we might need to understand the pull request -->
Fixes #10434 and #10802

sk- commented:
@patrickhulce when writing this change I was wondering what size should we show the report?

Right now we are showing the actual size which is multiplied by the original DPR, but that may be too confusing or annoying for the users.

Maybe we should just show the size we are using to decide whether the image is valid or not, and change the header title to Minimum accepted size.

Also note that with this change, the minimum effective DPR used is now 1.0, even for larger images, which shouldn't be a problem, as there's another audit checking the same.

patrickhulce commented on lighthouse-core/audits/image-size-responsive.js:
```diff
@@ -137,8 +140,9 @@ function allowedImageSize(displayedWidth, displayedHeight, DPR) {
   }
-  const width = Math.ceil(factor * DPR * displayedWidth);
-  const height = Math.ceil(factor * DPR * displayedHeight);
+  const allowedDpr = quantizeDpr(factor * quantizeDpr(DPR));
```
This seems like a major step backwards for large images from the situation today. I was thinking we quantize the DPR *and then also allow the factor*, not treat the factor as a modifier on the DPR. 

The reasons for factor and DPR quantization are somewhat orthogonal so mixing them confuses things a bit IMO.

I think we need both...
- DPR quantization, to support 2.625 and other non-standard DPR factors
- ~25% allowance, to support large fluidly sized responsive images that have discrete breakpoints that won't match the exact size.

sk- replied:
@patrick, this actually is an improvement in terms of being laxer at checking images.

Before for larger images it would use an effective DPR of 0.75* ORIGINAL_DPR, while with this PR it will use an effective DPR of floor (0.8 * floor(DPR)).

Below are the values 
```
Original DPR | Current Used DPR large | Used DPR PR small | Used DPR PR large
---------------------------------------------------------------------------------
1.0          | 0.75                   | 1.0               | 1.0
1.5          | 1.125                  | 1.5               | 1.0            
2.0          | 1.5                    | 2.0               | 1.5
2.625        | 1.96875                | 2.0               | 1.5         
3.0          | 2.25                   | 3.0               | 2.0
4.0          | 3.0                    | 4.0               | 3.0
5.0          | 3.75                   | 5.0               | 4.0
6.0          | 4.5                    | 6.0               | 4.0
``` 

Factor and quantization are not orthogonal, as they are basically a way of computing a minimum allowed dpr for both small and large images.

If you look at the table above, you can see that for integer DPRs the values before and after are pretty much the same, the difference only starts to be bigger at a DPR of 7, which there are no devices yet with such characteristics.

The rationale for quantizing twice was that the first quantization is to normalize non-standards ratios. While the second was to reduce the number of reports, as it's more likely to have an image which is proportional to any of the quantized DPRs, than having one proportional to the DPRs being used before. Take for example the example of a large image with a DPR of 3. Before we were requesting an effective DPR of 2.25, but that value is rather strange for users to supply. They either provided a 2x image or a 3x. If it is the former, then the check would fail, and if it is the latter, then the factor is not really doing any job.

I get that the implementation may now be complicated, so I thought about changing the strategy and have something like the following:

```ts
const KNOWN_DPRS = [1.0, 1.5, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0];
let allowedDpr = max(KNOWN_DPRS.filter(dpr => dpr <= DPR));
if (allowedDpr === undefined) {
  allowedDpr = DPR; // Or should it be 1.0?
}
if (largeImage) {
  // Here we allowed a lower dpr which is one level below. So 2.0 -> 1.5, 3.0 -> 2, etc
  allowedDpr = prevDpr(KNOWN_DPRS, allowedDpr);
}
```.


Also any thoughts regarding my comment about the reported size required?

patrickhulce replied:
Maybe I'm missing something, but your table shows the exact phenomenon I was talking about. The 1 DPR situation has regressed for large images.

It's great to demonstrate that other DPRs will have relax but 90+% of Lighthouse usage is the standard mobile and desktop profiles which use 2.625 and 1, respectively. I'm hesitant to become *more aggressive* with telling people to serve larger images in nearly half of the config usage we expect.

> Factor and quantization are not orthogonal, as they are basically a way of computing a minimum allowed dpr for both small and large images.

They are orthogonal from the perspective of *why* we're creating the allowances.

- We're creating the quantization of DPR allowance because it's unreasonable to expect developers to generate 2.625x versions of all their images.
- We're creating the factor because some responsive image implementations for fluid images require discrete breakpoints that won't always be exactly the displayed size, and it's OK if the images are scaled a bit.

Yes, I get that when you do the math it's just a single adjustment to make, but *why* we're creating the allowance is different. Expressing it all as "you must serve images at least X.xxx the exact displayed size when DPR is Y" is merging these two allowances in a way that masks the original reasoning behind them. It is not our intention that anyone attempt to create 1.5x images at every possible pixel size to serve to DPR 2 clients. We just want them to create DPR 2 and then give them an allowance in the *dimensions* of that DPR 2 image. I think this also opens the door for us to one day drop the dimension allowance if the image is a fixed size. Like you said "the second was to reduce the number of reports" and I'm just advocating expressing this as an adjustment to the required area rather than the required DPR :) 

It's a fairly minor point in the implementation as it really only amounts to a name change and removal of the second quantize call, but what I don't want lost is this reasoning we've established (for both end user documentation and for our future sanity when we come back to read this code years later wondering why it was done this way). IMO, splitting these factors in the code is an easy way to make this rationale clearer without enormous comments.

I think this comes down to just like a two line change. I don't think what you have right now requires major refactors.

```js
// add comment explaining two-pronged approach to function definition
const requiredDpr = quantizeDpr(DPR);
const width = Math.ceil(factor * requiredDpr * displayedWidth)
const height = Math.ceil(factor * requiredDpr * displayedHeight)
```

sk- replied:
> Maybe I'm missing something, but your table shows the exact phenomenon I was talking about. The 1 DPR situation has regressed for large images.
Ok, that's easy to fix in the quantization algorithm. Note though, that actually this is not a terrible change as there's already the audit `image-aspect-ratio`.

> 
> It's great to demonstrate that other DPRs will have relax but 90+% of Lighthouse usage is the standard mobile and desktop profiles which use 2.625 and 1, respectively. I'm hesitant to become _more aggressive_ with telling people to serve larger images in nearly half of the config usage we expect.
Changing the above code will fix this.

> 
> > Factor and quantization are not orthogonal, as they are basically a way of computing a minimum allowed dpr for both small and large images.
> 
> They are orthogonal from the perspective of _why_ we're creating the allowances.
> 
> * We're creating the quantization of DPR allowance because it's unreasonable to expect developers to generate 2.625x versions of all their images.
> * We're creating the factor because some responsive image implementations for fluid images require discrete breakpoints that won't always be exactly the displayed size, and it's OK if the images are scaled a bit.
My point, shown by the table, is that the factor by itself does not achieve the flexibility required to support discrete breakpoints.

Today if you have a DPR of 3, larger images require an image which is 2.25x, which strikes as an odd breakpoint to me. Instead with the proposed change, it will be just 2, which is much more reasonable.

I grant that there may be some difference on the DPR we use instead of the actual device DPR, and the breakpoints used, but that is quite blurry. For example if a common breakpoint is 2.5, then why not change the default 2.625 DPR to 2.5.
> 
> Yes, I get that when you do the math it's just a single adjustment to make, but _why_ we're creating the allowance is different. Expressing it all as "you must serve images at least X.xxx the exact displayed size when DPR is Y" is merging these two allowances in a way that masks the original reasoning behind them. It is not our intention that anyone attempt to create 1.5x images at every possible pixel size to serve to DPR 2 clients. We just want them to create DPR 2 and then give them an allowance in the _dimensions_ of that DPR 2 image. I think this also opens the door for us to one day drop the dimension allowance if the image is a fixed size. Like you said "the second was to reduce the number of reports" and I'm just advocating expressing this as an adjustment to the required area rather than the required DPR :)
I'm not sure I follow this comment. Do you mean you are ok, with the reported size as it is today, or should that incorporate at least one dpr quantization?
> 
> It's a fairly minor point in the implementation as it really only amounts to a name change and removal of the second quantize call, but what I don't want lost is this reasoning we've established (for both end user documentation and for our future sanity when we come back to read this code years later wondering why it was done this way). IMO, splitting these factors in the code is an easy way to make this rationale clearer without enormous comments.
As said above, I think the second quantize call is still needed. Otherwise, the factor is not having any real impact.
> 
> I think this comes down to just like a two line change. I don't think what you have right now requires major refactors.
> 
> ```js
> // add comment explaining two-pronged approach to function definition
> const requiredDpr = quantizeDpr(DPR);
> const width = Math.ceil(factor * requiredDpr * displayedWidth)
> const height = Math.ceil(factor * requiredDpr * displayedHeight)
> ```

patrickhulce replied:
> Note though, that actually this is not a terrible change as there's already the audit image-aspect-ratio.

I'm not sure how image-aspect-ratio applies here. I was worried about a 400x200 breakpoint being used for a 402x201 display size and Lighthouse fails.

> Today if you have a DPR of 3, larger images require an image which is 2.25x, which strikes as an odd breakpoint to me. Instead with the proposed change, it will be just 2, which is much more reasonable.

Perhaps there's some deeper disconnect here. I'm still confused by this framing of the effective DPR required. When I'm creating responsive images I don't think I've ever thought to myself what the effective DPR is and tried to reach some minimum there.

Instead I (or an algorithm from a build step) tend say oh look here's my source image. It's 2000x1000. I know it's going to be displayed at 400x200 or 600x300 depending on viewport, so I'll need 2x and 3x versions of those. 

> Otherwise, the factor is not having any real impact.

I'm not sure why you say no real impact. It shouldn't have any impact on the DPR the user targets (as intended) but it allows some scaling in large images, like we allow today before any of these changes.

> I'm not sure I follow this comment. Do you mean you are ok, with the reported size as it is today, or should that incorporate at least one dpr quantization?

I like reporting the expected size as we do today. It's immediately derivable from the displayed size and DPR and does not subject the user to any of this confusing logic. The fact that we're more lenient than that number in some cases is just a nice bonus for them :)

> Ok, that's easy to fix in the quantization algorithm.

I'm afraid we're talking past each other here because the *only* purpose of my very long comment was to try to and split the `factor` *out* from the concept of quantization. Even if you disagree with the rest of my reasoning, can we both agree that calling something "quantization" when it returns any decimal value less than 1 is a bit of a stretch? 😄

patrickhulce commented:
Somewhat unrelated but while we're improving this audit...

The docs link seems to 404 now and the title seems to be misleading some folks (https://github.com/GoogleChrome/lighthouse/issues/10802), maybe we could reword and find a new link?

I'm thinking something along the lines of....

`Serves images with appropriate resolution`
`Serves images with low resolution`
https://web.dev/serve-responsive-images/

perhaps?

sk- commented:
> Somewhat unrelated but while we're improving this audit...
> 
> The docs link seems to 404 now and the title seems to be misleading some folks (#10802), maybe we could reword and find a new link?
> 
> I'm thinking something along the lines of....
> 
> `Serves images with appropriate resolution`
> `Serves images with low resolution`
> https://web.dev/serve-responsive-images/
> 
> perhaps?

@patrickhulce Should I create another PR for this, or is it ok to change the titles in this one?

Regarding the url, the idea was to write a new article, based on both the one you linked and https://web.dev/image-aspect-ratio/.

Of course for now I think it's better to link to the one you provided instead of 404.

patrickhulce commented:
> @patrickhulce Should I create another PR for this, or is it ok to change the titles in this one?

Fine to change the titles here :)

> Regarding the url, the idea was to write a new article, based on both the one you linked and https://web.dev/image-aspect-ratio/.

Yes, eventually it should get its own, but in the meantime agreed that the other one is better than 404

sk- commented:
@patrickhulce I changed the PR to include the changes to the titles and link. Not sure if this would close #10802 or not.

I also reverted the change to quantization to truly be such a function. Also changed the PR to only apply the quantization once and keep the factor as it was.

patrickhulce commented:
@Beytoven did you also want to take a look?
